shell/src/Luna/Shell/GenerateDocumentation.hs

```diff
@@ -0,0 +1,99 @@
+import qualified Luna.Pass.Data.Stage          as TC
+import qualified Luna.Pass.Scheduler           as Scheduler
+import qualified Luna.Pass.Sourcing.Data.Class as Class
+import qualified Luna.Pass.Sourcing.Data.Def   as Def
+import qualified Luna.Pass.Sourcing.Data.Unit  as Unit
+import qualified Luna.Pass.Sourcing.UnitLoader as ModLoader
+import qualified Luna.Pass.Sourcing.UnitMapper as UnitMap
+import qualified Luna.Shell.CWD                as CWD
+import qualified Path
+import qualified System.Directory              as Directory
+
+import Data.Aeson                    (ToJSON)
+import Data.Map                      (Map)
+import Luna.Pass.Sourcing.Data.Class (Class)
+import Luna.Pass.Sourcing.Data.Unit  (Unit)
+import Luna.Pass.Sourcing.Data.Def   (Def, Documented (..))
+
+data DefDocumentation = DefDocumentation
+    { name :: Text
+    , documentation :: Maybe Text
+    } deriving (Show, Eq, Generic)
+
+data ClassDocumentation = ClassDocumentation
+    { name :: Text
+    , documentation :: Maybe Text
+    , methods :: [DefDocumentation]
+    } deriving (Show, Eq, Generic)
+
+data UnitDocumentation = UnitDocumentation
+    { name :: Text
+    , documentation :: Maybe Text
+    , classes   :: [ClassDocumentation]
+    , functions :: [DefDocumentation]
+    } deriving (Show, Eq, Generic)
+
+data ProjectDocumentation = ProjectDocumentation
+    { name :: Text
+    , documentation :: Maybe Text
+    , units :: [UnitDocumentation]
+    } deriving (Show, Eq, Generic)
+
+instance ToJSON DefDocumentation
+instance ToJSON ClassDocumentation
+instance ToJSON UnitDocumentation
+instance ToJSON ProjectDocumentation
+
+documentDef :: IR.Name -> Documented Def -> DefDocumentation
+documentDef name (Documented doc _) = DefDocumentation (convert name) doc
+
+documentClass :: IR.Name -> Documented Class -> ClassDocumentation
+documentClass name (Documented doc cls)
+    = ClassDocumentation (convert name) doc methodDocs where
+        methods    = Map.toList $ unwrap $ cls ^. Class.methods
+        methodDocs = uncurry documentDef <$> methods
+
+documentUnit :: Text -> Unit -> UnitDocumentation
+documentUnit name unit = UnitDocumentation name Nothing classDocs funDocs where
+    funs      = Map.toList $ unwrap $ unit ^. Unit.definitions
+    classes   = Map.toList $ unit ^. Unit.classes
+    funDocs   = uncurry documentDef   <$> funs
+    classDocs = uncurry documentClass <$> classes
+
+defaultOutFileName :: FilePath
+defaultOutFileName = "documentation.json"
+
+generateDocumentation :: FilePath -> FilePath -> IO ()
+generateDocumentation outFile' modPath' = do
+    let outFile = if null outFile' then defaultOutFileName else outFile'
+    modPath      <- if null modPath' then CWD.get else pure modPath'
+    canonModPath <- Directory.canonicalizePath modPath
+    path         <- Path.parseAbsDir canonModPath
+    sourcesMap   <- fmap Path.toFilePath . Bimap.toMapR
+                        <$> Package.findPackageSources path
```

Comment:
I'd definitely go here this way:

```haskell
    sourcesMap   <- fmap Path.toFilePath . Bimap.toMapR
                <$> Package.findPackageSources path
```

Reply:
This breaks `4*k` space indent rule. We've agreed to not do it because change of `sourcesMap` name will make it unaligned and it is super easy to forget about fixing it.